### PR TITLE
refactor: Replace `FrechetFrame` with `DeformedFrame`/`MatrixLogFrame` and add `Frame.structure` dispatch

### PR DESCRIPTION
**Pull request by n-gao**
Replaces `FrechetFrame` with a composable `DeformedFrame` + `MatrixLogFrame` design, and adds a `structure` property to the `Frame` protocol so that geometric primitives (inverse, volume, coordinate transforms) dispatch to the appropriate fast path.

**`DeformedFrame`** separates the fixed reference (`base`, stop-gradient) from the differentiable `deformation` frame, rather than hard-coding a lower-triangular `tril` array and a scalar `cell_factor`. The deformation type is now a constructor argument: passing `LogTriclinicFrame` (default) reproduces the previous ASE `FrechetCellFilter` behaviour with a `(6,)` lower-triangular parameterisation; passing `MatrixLogFrame` gives a full `3×3` exponential-map deformation that supports rotated/sheared cells and keeps every leaf `(..., 3, 3)` so per-system batched slicing works uniformly.

**`MatrixLogFrame`** stores a full `3×3` generator `A` and computes `vectors = expm(A / cell_factor)` via `jax.scipy.linalg.expm`. Its VJP is the `expm_frechet` adjoint, so `parameter_gradient` / `vectors_gradient` are exact with no custom rule. Construction from a lower-triangular reference uses the exact `triangular_3x3_logm` to avoid numerical issues with repeated eigenvalues (cubic/tetragonal cells).

**`structure` property** is added to the `Frame` protocol with three values — `LOWER_TRIANGULAR`, `DIAGONAL`, and `GENERAL` — and implemented on every concrete frame. `BaseFrame` uses it to route `inverse_vectors`, `volume`, `to_fractional`, `to_real`, and `materialize` through `structured_det_and_inverse` / `structured_matmul` instead of the previously hard-coded triangular helpers. `MaterializedFrame` stores and propagates the structure flag so materialized frames keep the fast path of their source. `DeformedFrame.structure` is derived from the product of its base and deformation structures via `matmul_structure`.

**`cell_factor` handling** is generalised from a static scalar to a per-system array in both `LogTriclinicFrame` and `MatrixLogFrame`. It is broadcast to the full parameter shape (sharing the leading batch dimension) and stored as a stop-gradient leaf, with a `.factor` property exposing the per-system scalar. `relax_state_from_ase` now computes `cell_factor` as a per-system atom count via `bincount` over system indices rather than a single global count.

**`Edges.absolute_shifts`** is updated to call `cell.materialize().to_real(shifts)` instead of the hard-coded `triangular_3x3_matmul`, so it correctly handles general (non-triangular) cells.

### CODE REVIEW

**Comment by n-gao** [comment]
* **#213** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/213?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2  dependent PRs ([#215](https://github.com/cusp-ai-oss/kUPS/pull/215) <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/215?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#216](https://github.com/cusp-ai-oss/kUPS/pull/216) <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/216?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#212** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/212?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#211** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/211?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#210** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/210?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#204** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/204?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#209** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/209?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#203** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/203?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#202** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/202?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/202?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#201** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/201?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#186** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/186?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#185** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/185?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#183** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/183?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#182** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/182?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#181** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/181?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#180** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/180?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#178** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/178?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#171** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/171?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#173](https://github.com/cusp-ai-oss/kUPS/pull/173) <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/173?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#170** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/170?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#206** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/206?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#205** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/205?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Review by felixmusil — APPROVED** [review]

**Comment by n-gao** [comment]
### Merge activity

* **Jul 1, 3:15 PM UTC**: @n-gao merged this pull request with [Graphite](https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/202).
